### PR TITLE
Fix slow hierarchy filtering with large number of filtered paths

### PR DESCRIPTION
**Pull request by JonasDov**
Closes [#706](https://github.com/iTwin/presentation/issues/706)

### CODE REVIEW

File: apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts [hunk 1]
```diff
@@ -214,7 +214,6 @@ describe("Hierarchies", () => {
               children: [
                 NodeValidators.createForGenericNode({
                   key: "custom2",
-                  autoExpand: false,
```
Comment (grigasp) [inline apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts:217]: why was this removed?

File: apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts [hunk 2]
```diff
@@ -1528,19 +1531,24 @@ describe("Hierarchies", () => {
...
+                  const identifierIndexAdded = identifierIndex + 1;
                   return (
-                    path.length &&
-                    HierarchyNodeIdentifier.isGenericNodeIdentifier(path[0]) &&
-                    path[0].source === "custom-provider" &&
-                    path[0].id === myNode.key.id
+                    path.length > identifierIndexAdded &&
+                    HierarchyNodeIdentifier.isGenericNodeIdentifier(path[identifierIndexAdded]) &&
+                    path[identifierIndexAdded].source === "custom-provider" &&
+                    path[identifierIndexAdded].id === myNode.key.id
```
Comment (grigasp) [inline apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts:1551]: Looks like implementing hierarchy filtering in a custom hierarchy provider became much more complicated. We should not enforce any specific implementation requirements for custom providers - they should be able to choose the most convenient way to filter the hierarchy.

File: packages/hierarchies/api/presentation-hierarchies.api.md [hunk 3]
```diff
@@ -159,7 +159,8 @@ interface ECSqlValueSelector {
 
 // @beta
 export function extractFilteringProps(rootLevelFilteringProps: HierarchyFilteringPath[], parentNode: Pick<NonGroupingHierarchyNode, "filtering"> | undefined): {
-    filteredNodePaths: HierarchyFilteringPath[];
+    filterPathsIdentifierPositions?: Array<[number, number]>;
+    nodeIdentifierPaths: HierarchyFilteringPath[];
```
Comment (grigasp) [inline packages/hierarchies/api/presentation-hierarchies.api.md:163]: This is a breaking change that we can't make anymore

Comment (grigasp) [reply packages/hierarchies/api/presentation-hierarchies.api.md:163]: The other changes in this file are also breaking

File: apps/performance-tests/src/hierarchies/Filtering.test.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  for (let i = 0; i < totalNumberOfFilteringPaths; ++i) {
+    const hundredsPosition = Math.floor(i / itemsPerGroup);
+    const id = `0x${(i + startingIndex).toString(16)}`;
+    const nearestParentId = `0x${parentIdsArr[i % Math.ceil(totalNumberOfFilteringPaths / maximumNumberOfPathsForSingleParent)].toString(16)}`;
+    filteringPaths.push([
+      { className: `${schemaName}.${defaultClassName}_0`, id: `0x${startingIndex.toString(16)}` },
+      { className: `${schemaName}.${defaultClassName}_0`, id: nearestParentId },
+      { className: `${schemaName}.${defaultClassName}_${hundredsPosition}`, id },
+    ]);
+  }
```
Comment (grigasp) [inline apps/performance-tests/src/hierarchies/Filtering.test.ts:156]: please add some comments explaining what's happening here

**Review by grigasp — COMMENTED** [review]
didn't look at the changes in `FilteringHierarchyDefinition` - need to resolve breaking changes first

File: apps/performance-tests/src/hierarchies/Filtering.test.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import { createBisInstanceLabelSelectClauseFactory } from "@itwin/presentation-shared";
+
+describe("filtering", () => {
+  const totalNumberOfFilteringPaths = 10000;
+  const startingIndex = 20;
+  const maximumNumberOfPathsForSingleParent = 500;
+  const parentIdsArr = new Array<number>();
+  for (let i = maximumNumberOfPathsForSingleParent, j = startingIndex + 1; i <= totalNumberOfFilteringPaths; i += maximumNumberOfPathsForSingleParent, ++j) {
+    parentIdsArr.push(j);
+  }
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply apps/performance-tests/src/hierarchies/Filtering.test.ts]: Moved

File: apps/performance-tests/src/hierarchies/Filtering.test.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                        extendedData: {
+                          ecInstanceId: { selector: `this.ECInstanceId` },
+                        },
+                      })}
+                      FROM ${fullClassName} AS this
+                      WHERE this.ECInstanceId = ${startingIndex}
+                    `,
+                  },
+                },
+              ];
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply apps/performance-tests/src/hierarchies/Filtering.test.ts]: Yes, added `createHierarchyLevelDefinition` function

File: apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts (hunk 1 above)
Comment (JonasDov) [reply apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts]: Added back

File: packages/hierarchies/api/presentation-hierarchies.api.md (hunk 3 above)
Comment (JonasDov) [reply packages/hierarchies/api/presentation-hierarchies.api.md]: Changed

File: apps/performance-tests/src/hierarchies/Filtering.test.ts (hunk 4 above)
Comment (JonasDov) [reply apps/performance-tests/src/hierarchies/Filtering.test.ts]: Didn't add a comment, just simplified it.

File: .changeset/tall-comics-compare.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+"@itwin/presentation-hierarchies": minor
+---
+
+Increased the speed of hierarchy filtering with large number of filtered paths.
```
Comment (grigasp) [inline .changeset/tall-comics-compare.md:5]: should list all the api additions / changes you made in this pr

File: apps/performance-tests/src/hierarchies/Filtering.test.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            { className: `${schemaName}.${defaultClassName}_0`, id: `0x${physicalElementsSmallestIndex.toString(16)}` },
+            { className: `${schemaName}.${defaultClassName}_${i / itemsPerGroup}`, id: `0x${(i + physicalElementsSmallestIndex).toString(16)}` },
+            { className: `${schemaName}.${defaultClassName}_${j / itemsPerGroup}`, id: `0x${(j + physicalElementsSmallestIndex).toString(16)}` },
+          ]);
+        }
+      }
+
+      const iModel = SnapshotDb.openFile(Datasets.getIModelPath("50k flat elements"));
+      const fullClassName = PhysicalElement.classFullName.replace(":", ".");
+      const createHierarchyLevelDefinition = async (imodelAccess: ECSchemaProvider & ECClassHierarchyInspector, eCInstanceIdCondition: string) => {
```
Comment (grigasp) [inline apps/performance-tests/src/hierarchies/Filtering.test.ts:40]: - letter casing problem in `eCInstanceIdCondition`
- I suggest changing it to `whereClause: (alias: string) => string`

File: apps/performance-tests/src/hierarchies/Filtering.test.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            //
+            // We need to split the hierarchy in 100 parts, because we are using 50000 paths and there is a limit of 500 filtering paths for a single parent.
+
+            if (!props.parentNode) {
+              return createHierarchyLevelDefinition(imodelAccess, `= ${physicalElementsSmallestIndex}`);
+            }
+
+            if (props.parentNode?.extendedData?.ecInstanceId === physicalElementsSmallestIndex) {
+              return createHierarchyLevelDefinition(imodelAccess, `IN (${parentIdsArr.join(", ")})`);
+            }
```
Comment (grigasp) [inline apps/performance-tests/src/hierarchies/Filtering.test.ts:86]: Why use extended data for this? We could do this instead:
```ts
if (parentNode && HierarchyNode.isInstancesNode(parentNode) && parentNode.key.instanceKeys.some(({ id }) => Id64.getLocalId(id) === physicalElementsSmallestIndex)) {
  // ...
}
```

File: apps/performance-tests/src/hierarchies/Filtering.test.ts [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import { IModelDb, PhysicalElement, SnapshotDb } from "@itwin/core-backend";
+import { createNodesQueryClauseFactory, HierarchyFilteringPath } from "@itwin/presentation-hierarchies";
+import { Datasets } from "../util/Datasets";
+import { run } from "../util/TestUtilities";
+import { ProviderOptions, StatelessHierarchyProvider } from "./StatelessHierarchyProvider";
+import { createBisInstanceLabelSelectClauseFactory, ECClassHierarchyInspector, ECSchemaProvider } from "@itwin/presentation-shared";
+
+describe("filtering", () => {
+  const totalNumberOfFilteringPaths = 50000;
+  const physicalElementsSmallestIndex = 20;
```
Comment (grigasp) [inline apps/performance-tests/src/hierarchies/Filtering.test.ts:16]: I believe this is not an index, but an id in decimal form.

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 11]
```diff
@@ -21,6 +21,7 @@ import {
 } from "./IModelHierarchyDefinition";
 import { ProcessedGroupingHierarchyNode, ProcessedHierarchyNode, SourceHierarchyNode, SourceInstanceHierarchyNode } from "./IModelHierarchyNode";
 import { defaultNodesParser } from "./TreeNodesReader";
+import { Dictionary } from "@itwin/core-bentley";
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:24]: Imports from other packages should be listed before imports from local modules. I suggest using https://marketplace.visualstudio.com/items?itemName=mike-co.import-sorter for that.

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 12]
```diff
@@ -151,7 +239,7 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
             this._imodelAccess,
             (def, matchingFilters) => {
               const filteredChildrenIdentifierPaths = matchingFilters.reduce(
-                (r, c) => [...r, ...c.childrenIdentifierPaths],
+                (r, c) => [...r, ...(c.childrenIdentifierPaths ? c.childrenIdentifierPaths : [])],
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:242]:
```suggestion
                (r, c) => [...r, ...(c.childrenIdentifierPaths ?? [])],
```

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 13]
```diff
@@ -219,14 +307,18 @@ async function matchFilters<
...
           isFilterTarget: false,
         };
         matchingFilters.push(entry);
       }
-      const remainingPath = path.slice(1);
-      if (remainingPath.length > 0) {
-        entry.childrenIdentifierPaths.push(options ? { path: remainingPath, options } : remainingPath);
+
+      if (path.length > 1) {
+        if (HierarchyNodeIdentifier.isGenericNodeIdentifier(nodeId)) {
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:316]: why do we only need to do this for generic nodes from now on?

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 14]
```diff
@@ -323,12 +419,12 @@ export function applyECInstanceIdsFilter(
...
           ${matchingFilters
             .map((mc) =>
               mc.isFilterTarget
-                ? `VALUES (${mc.id.id}, 1, ${mc.filterTargetOptions ? `'${JSON.stringify(mc.filterTargetOptions)}'` : "CAST(NULL AS TEXT)"}, '${JSON.stringify(mc.childrenIdentifierPaths)}')`
-                : `VALUES (${mc.id.id}, 0, CAST(NULL AS TEXT), '${JSON.stringify(mc.childrenIdentifierPaths)}')`,
+                ? `VALUES (${mc.id.id}, 1, ${mc.filterTargetOptions ? `'${JSON.stringify(mc.filterTargetOptions)}'` : "CAST(NULL AS TEXT)"}, '${mc.id.className}')`
+                : `VALUES (${mc.id.id}, 0, CAST(NULL AS TEXT), '${mc.id.className}')`,
             )
             .join(" UNION ALL ")}
         )`,
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:430]: I'm now thinking that we could group all instance keys by class name and change the CTE from 

```sql
VALUES(...) UNION ALL VALUES(...) UNION ALL ...
```
to 
```sql
SELECT ... FROM ONLY {class name 1} WHERE ECInstanceId IN (?,?,?,...)
UNION ALL
SELECT ... FROM ONLY {class name 2} WHERE ECInstanceId IN (?,?,?,...)
UNION ALL 
...
```

This should further reduce the size of SQL statement and possibly even help for https://github.com/iTwin/presentation/issues/722

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 15]
```diff
@@ -106,16 +152,58 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
         : undefined;
-      const filteredChildrenIdentifierPaths: HierarchyFilteringPath[] | undefined = row[ECSQL_COLUMN_NAME_FilteredChildrenPaths]
-        ? JSON.parse(row[ECSQL_COLUMN_NAME_FilteredChildrenPaths])
-        : undefined;
-      const defaultNode = (this._source.parseNode ?? defaultNodesParser)(row);
+
+      // istanbul ignore next
+      const allFilterPathsIdentifierPositions: Array<[number, number]> | undefined = row[ECSQL_COLUMN_NAME_FilterECInstanceId]
+        ? (this._pathsIdentifierPositions.get({
+            id: `0x${row[ECSQL_COLUMN_NAME_FilterECInstanceId].toString(16)}`,
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:165]: we should select the id as id64 string, otherwise there's a risk for it to exceed javascript's number

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 16]
```diff
@@ -106,16 +152,58 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+      for (const [pathIndex, identifierIndex] of allFilterPathsIdentifierPositions) {
+        // istanbul ignore if
+        if (this._nodeIdentifierPaths.length <= pathIndex) {
+          continue;
+        }
+        const path = HierarchyFilteringPath.normalize(this._nodeIdentifierPaths[pathIndex]);
+        // istanbul ignore if
+        if (path.path.length <= identifierIndex) {
+          continue;
+        }
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:184]: I'd like tests to cover these two edge cases.

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 17]
```diff
@@ -106,16 +152,58 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
       const isFilterTarget: boolean = !!row[ECSQL_COLUMN_NAME_IsFilterTarget];
       const filterTargetOptions: HierarchyFilteringPathOptions | undefined = row[ECSQL_COLUMN_NAME_FilterTargetOptions]
         ? JSON.parse(row[ECSQL_COLUMN_NAME_FilterTargetOptions])
         : undefined;
-      const filteredChildrenIdentifierPaths: HierarchyFilteringPath[] | undefined = row[ECSQL_COLUMN_NAME_FilteredChildrenPaths]
-        ? JSON.parse(row[ECSQL_COLUMN_NAME_FilteredChildrenPaths])
-        : undefined;
-      const defaultNode = (this._source.parseNode ?? defaultNodesParser)(row);
+
+      // istanbul ignore next
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:162]: There's a lot of code for doing one thing - it makes `parseNode` hard to read. Could we move it into a separate function?

File: .changeset/tall-comics-compare.md (hunk 7 above)
Comment (JonasDov) [reply .changeset/tall-comics-compare.md]: Added

File: apps/performance-tests/src/hierarchies/Filtering.test.ts (hunk 8 above)
Comment (JonasDov) [reply apps/performance-tests/src/hierarchies/Filtering.test.ts]: Changed

File: apps/performance-tests/src/hierarchies/Filtering.test.ts (hunk 9 above)
Comment (JonasDov) [reply apps/performance-tests/src/hierarchies/Filtering.test.ts]: Removed extendedData

File: apps/performance-tests/src/hierarchies/Filtering.test.ts (hunk 10 above)
Comment (JonasDov) [reply apps/performance-tests/src/hierarchies/Filtering.test.ts]: Renamed

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 11 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: I had opened presentation repo in co-space, for some reason import sorter does not work in co-space. Fixed the order

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 13 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Because we are not using `childrenIdentifierPaths` for `InstanceNodeIdentifiers`. Changed the implementation, now `matchFilters` function takes in another argument: `extractChildrenIdentifierPaths` which determines whether or not to retrieve `childrenIdentifierPaths`

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 14 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:433]: I don't think this suggestion would work as keys with the same className and different id's can have different `IsFilterTarget` and `FilterTargetOptions` values.

Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:433]: Keeping as is

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 15 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Changed the query to return id64 string instead.

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 16 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: These two edge cases should never happen because:
We create `_pathsIdentifierPositions` array based on `_nodeIdentifierPaths`. So `_pathsIdentifierPositions` always has indexes pointing to an existing place in the `_nodeIdentifierPaths` array.
These two checks were added just a precaution. Removed them.

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 17 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Addded `getChildrenIdentifierPathsFromIdentifier` function.

File: .changeset/tall-comics-compare.md [hunk 18]
```diff
@@ -0,0 +1,9 @@
+---
+"@itwin/presentation-hierarchies": minor
+---
+
+- Increased the speed of hierarchy filtering with large number of filtered paths.
+
+- Added `compare` function to `HierarchyNodeIdentifier`.
+
+- Changed `NodeParser`, it now returns `SourceInstanceHierarchyNode | Promise<SourceInstanceHierarchyNode>`.
```
Comment (grigasp) [inline .changeset/tall-comics-compare.md:9]: The first row should not be a bullet point, otherwise the final changelog renders weirdly. Also made API addition descriptions more specific.

```suggestion
Increased the speed of hierarchy filtering with large number of filtered paths.

- Added `HierarchyNodeIdentifier.compare` function to compare two identifiers.
- Changed `NodeParser` (return type of `HierarchyDefinition.parseNode`), to additionally allow returning a promise, so instead of just `SourceInstanceHierarchyNode` it can now also return `Promise<SourceInstanceHierarchyNode>`.
```

**Review by grigasp — COMMENTED** [review]
I'd like a few tests to be added to `FilteringHierarchyDefinition.test.ts` to cover the edge cases we discussed earlier:
- filtering by multiple paths where ids are the same, but classes are different,
- filtering by multiple paths where the same identifier is at different positions of the path.

**Comment by grigasp** [comment]
any idea why "Hierarchies performance benchmark" comment now contains unified selection performance numbers?

**Comment by JonasDov** [comment]
> FilteringHierarchyDefinition.test.ts

Added 3 tests:
- `sets correct filteredChildrenIdentifierPaths when nodes have same id's and different classNames that don't derive from one another`
- `sets correct filteredChildrenIdentifierPaths when same identifier is in different positions of the path`
- `sets correct filteredChildrenIdentifierPaths when nodes have same id's and different classNames that derive from one another`

**Comment by JonasDov** [comment]
> any idea why "Hierarchies performance benchmark" comment now contains unified selection performance numbers?

This was due to adding `spec: ["./lib/**/*.test.js"]` to `.mocharc.json` file.
I tried a few things to fix this:
1. Tried specifying `--spec` option in package.json, but it did not work. It seems from mocha docs, that they merge options that can be merged: https://mochajs.org/#merging
2. Tried using `MOCHA_OPTIONS`, but it did not work.
3. Tried to find a way to split  `.mocharc.json` file into two, but did not find a way to do so.
Decided to revert the change, and add the path to `launch.json`, by doing it this way the debugger works.

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 19]
```diff
@@ -86,6 +102,70 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+        }
+        ++similarityLength;
+        continue;
+      }
+      // istanbul ignore next
+      break;
+    }
+    entry = similarityLength;
+    this._pathsSimilarityLengthCache.set({ lhsIndex, rhsIndex }, entry);
+    return entry;
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:166]: what's the point of assigning this to `entry`?

```suggestion
    this._pathsSimilarityLengthCache.set({ lhsIndex, rhsIndex }, similarityLength);
    return similarityLength;
```

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 20]
```diff
@@ -188,12 +276,30 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+      if (identifierIndex !== validPath.path.length - identifiersCountAfter - 1) {
+        continue;
+      }
+
+      if (pathIndex !== validPathIndex) {
+        const similarityLength = await this.getPathsSimilarityLength(pathIndex, validPathIndex);
+        // istanbul ignore if
+        if (similarityLength < identifierIndex) {
+          continue;
+        }
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:298]: should be covered with tests

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 21]
```diff
@@ -86,6 +102,70 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+      }
+
+      // istanbul ignore next
+      if (HierarchyNodeIdentifier.isGenericNodeIdentifier(lhsIdentifier) && HierarchyNodeIdentifier.isGenericNodeIdentifier(rhsIdentifier)) {
+        if (lhsIdentifier.source !== rhsIdentifier.source || lhsIdentifier.id !== rhsIdentifier.id) {
+          break;
+        }
+        ++similarityLength;
+        continue;
+      }
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:160]: should be covered with tests

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 22]
```diff
@@ -86,6 +102,70 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+      // istanbul ignore next
+      break;
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:162]: should be covered with tests

File: packages/hierarchies/src/hierarchies/HierarchyNode.ts [hunk 23]
```diff
@@ -26,6 +26,8 @@ export type HierarchyNodeFilteringProps = {
   hasFilterTargetAncestor?: boolean;
   /** Paths to node's children that are filter targets. */
   filteredChildrenIdentifierPaths?: HierarchyFilteringPath[];
+  /** Paths index in all filter paths */
+  filteredChildrenIdentifierPathsIndex?: number[];
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/HierarchyNode.ts:30]: I don't understand the meaning of this. Can you explain in more detail?

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 24]
```diff
@@ -35,12 +35,28 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+   * This dictionary would return: 3, since both paths start with a, b, c identifiers (which are similar)
+   */
+  private _pathsSimilarityLengthCache: Dictionary<{ lhsIndex: number; rhsIndex: number }, number>;
 
   public constructor(props: FilteringHierarchyDefinitionProps) {
     this._imodelAccess = props.imodelAccess;
     this._source = props.source;
     this._nodeIdentifierPaths = props.nodeIdentifierPaths;
     this._pathsIdentifierPositions = this.createPathsIdentifierPositions(props.nodeIdentifierPaths);
+    this._pathsSimilarityLengthCache = this.createPathsSimilarityLengthCache();
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: At this point we have class 3 members + a ton of code for handling those paths. Maybe it would be worth moving all of that into a separate class responsible just for that?

File: apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts [hunk 25]
```diff
@@ -223,6 +223,106 @@ describe("Hierarchies", () => {
...
+              fullClassName: subjectClassName,
+              query: {
+                ecsql: `
+                  SELECT ${await selectQueryFactory.createSelectClause({
+                    ecClassId: { selector: `this.ECClassId` },
+                    ecInstanceId: { selector: `this.ECInstanceId` },
+                    nodeLabel: { selector: `this.CodeValue` },
+                    extendedData: {
+                      level: extendedDataLevel,
+                    },
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts]: Removed

File: apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts [hunk 26]
```diff
@@ -223,6 +223,106 @@ describe("Hierarchies", () => {
...
+            }
+            if (HierarchyNode.isInstancesNode(parentNode) && parentNode.label === "test subject 4") {
+              return createHierarchyLevelDefinition("3", (alias: string) => `WHERE ${alias}.Parent.Id = ${keys.rootSubject.id}`);
+            }
+            if (HierarchyNode.isInstancesNode(parentNode) && parentNode.label === "test subject 1" && parentNode.extendedData?.level === "3") {
+              return createHierarchyLevelDefinition("4", (alias: string) => `WHERE ${alias}.Parent.Id = ${keys.childSubject1.id}`);
+            }
+            return [];
+          },
+        };
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts]: Applied changes

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 27]
```diff
@@ -86,6 +102,70 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
...
+      }
+      if (lhs.lhsIndex !== rhs.lhsIndex) {
+        // istanbul ignore next
+        return lhs.lhsIndex > rhs.lhsIndex ? 1 : -1;
+      }
+      // istanbul ignore next
+      return lhs.rhsIndex > rhs.rhsIndex ? 1 : -1;
+    };
+    return new Dictionary<{ lhsIndex: number; rhsIndex: number }, number>(compare);
+  }
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Moved outside of the class

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 20 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: It is now covered

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 21 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Covered

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 22 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Covered

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 24 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Created a separate class for handling positions

File: packages/hierarchies/src/hierarchies/HierarchyNode.ts (hunk 23 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/HierarchyNode.ts:30]: This tells which original nodeIdentifierPath is filteredChildrenIdentifierPaths pointing to.
For example if we create a `FilteringHierarchyDefinition` with these `nodeIdentifierPaths`: 
```ts
nodeIdentifierPaths: [
  [a, b, c],
  [a, c, d],
  [a, b, d]
]
```
And try to get `filteredChildrenIdentifierPaths` for node b, 
The following `filteredChildrenIdentifierPaths` would be created:
```ts
filteredChildrenIdentifierPaths: [
  [c],
  [d]
]
```
And the following `filteredChildrenIdentifierPathsIndex` would be created:
```ts
filteredChildrenIdentifierPathsIndex: [0, 2]
```
filteredChildrenIdentifierPathsIndex tells us that first path ([c]) in `filteredChildrenIdentifierPaths` is created from 0th path ([a, b, c]) of `nodeIdentifierPaths`. And the the second path ([d]) is created from 2nd path ([a, b, d]).

Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/HierarchyNode.ts]: After talking in private, changed the implementation, this is no longer needed.

File: apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (grigasp) [reply apps/full-stack-tests/src/hierarchies/HierarchyFiltering.test.ts]: shouldn't the root "test subject 1" node have 2 children - "test subject 2" and "test subject 3"?

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 28]
```diff
@@ -199,10 +246,11 @@ async function matchFilters<
   TDefinition = TIdentifier extends InstanceKey ? InstanceNodesQueryDefinition : GenericHierarchyNodeDefinition,
 >(
   definition: TDefinition,
-  filteringProps: { filteredNodePaths: HierarchyFilteringPath[]; hasFilterTargetAncestor?: boolean },
+  filteringProps: { filteredNodePaths: HierarchyFilteringPath[]; hasFilterTargetAncestor?: boolean; filteredNodePathsIndex?: number[] },
```
Comment (grigasp) [inline packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:249]: is `filteredNodePathsIndex` needed anymore?

File: .changeset/tall-comics-compare.md (hunk 7 above)
Comment (grigasp) [reply .changeset/tall-comics-compare.md:5]: I'm now thinking it would be nice to also specify the scope of improvement, e.g.:

```
Increased the speed of hierarchy filtering with large number of filtered paths.

| Amount of paths | Before the change | After the change |
| --- | --- | --- |
| 1k | ...
| 10k | ...
| 50k | ...

In addition, changed `NodeParser` (return type of `HierarchyDefinition.parseNode`):
- It now can return a promise, so instead of just `SourceInstanceHierarchyNode` it can now also return `Promise<SourceInstanceHierarchyNode>`.
- Additionally, it now accepts an optional `parentNode` argument of `HierarchyDefinitionParentNode` type.
```

**Review by grigasp — APPROVED** [review]
Looks great now!

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts (hunk 28 above)
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: No, removed

File: .changeset/tall-comics-compare.md (hunk 7 above)
Comment (JonasDov) [reply .changeset/tall-comics-compare.md:5]: Updated
